layers/+lang/factor/funcs.el

```diff
@@ -8,11 +8,116 @@
 ;; This file is not part of GNU Emacs.
 ;;
 ;;; License: GPLv3
+(require 'loadhist)
+
+(defvar factor/ui-listener-root nil
+  "If set, commands that start the graphical listener will use
+this path to lok for the factor executable and fuel mode
+implementation. Intended to be customized per-project.")
+
+(defvar factor/ui-listener-args ""
+  "Extra arguments to the factor VM binary.")
 
 (defun factor//fuel-stack-effect ()
   "Small wrapper around factors stack effect help. If region is
-  active, use that, otherwise use sexp under point."
+active, use that, otherwise use sexp under point."
   (interactive)
   (if (region-active-p)
       (call-interactively 'fuel-stack-effect-region)
     (call-interactively 'fuel-stack-effect-sexp)))
+
+(defun factor//fuel-elisp-dir ()
+  "Get the directory of the currently loaded fuel implementation."
+  (expand-file-name (concat (feature-file 'factor-mode)
+                           "/..")))
+
+(defun factor//fuel-feature-p (fuel-directory feature)
+  "Check if FEATURE is provided by a file under FUEL-DIRECTORY."
+  (string-prefix-p fuel-directory (feature-file feature)))
+
+(defun factor//loaded-fuel-features ()
+  "Recursively determine all loaded features that belong to the
+currently loaded fuel implementation."
+  (cl-loop with fuel-directory = (factor//fuel-elisp-dir)
+           for fuel-features = '(fuel-mode factor-mode)
+           then (cl-union fuel-features new-features)
+           for next-features = (cl-remove-duplicates
+                                (cl-reduce 'cl-union
+                                           (mapcar (lambda(f) (file-requires (feature-file f)))
+                                                   fuel-features)))
+           for new-features = (cl-remove-if-not
+                               (lambda(f) (and (featurep f)
+                                               (factor//fuel-feature-p fuel-directory f)))
+                               (cl-set-difference next-features fuel-features))
+           while new-features
+           finally (return fuel-features)))
+
+(defun factor//disconnect-fuel ()
+  "Stop the current fuel connection if applicable."
+  (when (and (featurep 'fuel-listener)
+             fuel-listener--buffer)
+    (kill-buffer fuel-listener--buffer)))
+
+(defun factor//unload-fuel ()
+  "Close fuel connection and unload fuel code."
+  (factor//disconnect-fuel)
+  (cl-loop for f in (factor//loaded-fuel-features) do
+           (unload-feature f t)))
+
+(defun factor//load-fuel-from-path (path)
+  (let ((load-path (cons path load-path)))
+    (require 'fuel-mode)
+    (require 'factor-mode)))
+
+(defvar factor//ui-listener-process nil
```

Comment:
@smile13241324 Even though it is not a configuration variable, but a state variable used by private functions only?

Reply:
Yep, you need to mark your internal variables with the proper prefix, I think it is "--" but please check the conventions.

Reply:
Ok, will do that.  It still feels strange because `config.el` is loaded after `funcs.el`, but contains non-user-visible state exclusively used in `funcs.el`, thus breaking file consistency.  From `LAYERS.org`:

> This file configure the layer like declaring layer variables default values and setup some other variables related to the layer.

I understand the concept of layer variables to be something that can be set by the user from `.spacemacs`.  So to me, it is actually unexpected to find internal implementation details in `config.el`.

Reply:
Hmmm this is also not done consistently in the layers ... let me ask the other maintainers what's the official convention here.

Reply:
Not sure when I will get an answer, for now you can leave them in funcs.el, but please move them to the top of the file then or in a function.

However please change the name to use "--" so that it is clear that it is an internal variable.

Reply:
Will do.